File: iloadins/ins_file_loader_alt.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2026 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// On success: returns 0.
// On error: returns -1 and error description in `error'.
int8_t ins_file_loader_alt (temp_instrument_t *dst, const String *_fname, char **error)
{
  int8_t result = -1; // return value
  FILE *f = NULL;
  long fsize;
  tINS_DATA buffer;
  char fname[255+1];

  DBG_ENTER ("ins_file_loader_alt");

  StringToStr (fname, _fname, sizeof (fname) - 1);
  if ((f = fopen (fname, "rb")) == NULL) goto _err_fopen;

  if (fseek (f, 0, SEEK_END) != 0) goto _err_fread;
  if ((fsize = ftell (f)) < 0) goto _err_fread;
  if (fsize < (long)sizeof (buffer)) goto _err_format;
  if (fseek (f, 0, SEEK_SET) != 0) goto _err_fread;

  if (fread (&buffer, fsize, 1, f) == 0) goto _err_fread;

  dst->four_op = false;
  dst->use_macro = false;

  // instrument data
  memset (&dst->ins1.fm, 0, sizeof (dst->ins1.fm));
  switch (force_ins)
  {
    case 0:
      if (fsize == 12)
        import_standard_instrument_alt (&dst->ins1.fm, &buffer.idata);
      if (   (fsize == 12)
          && (   (buffer.idata.WAVEFORM_modulator > 3)
              || (buffer.idata.WAVEFORM_carrier > 3)
              || (buffer.idata.FEEDBACK_FM > 15)))
        import_hsc_instrument_alt (&dst->ins1.fm, &buffer.idata);
      else if (fsize > 12)
        import_sat_instrument_alt (&dst->ins1.fm, &buffer.idata);
      break;

    case 1: import_hsc_instrument_alt (&dst->ins1.fm, &buffer.idata); break;
    case 2: import_sat_instrument_alt (&dst->ins1.fm, &buffer.idata); break;
    case 3: import_standard_instrument_alt (&dst->ins1.fm, &buffer.idata); break;
    default: break;
  }

  // instrument name
  SetLength (dst->ins1.name, 0);
  set_default_ins_name_if_needed (dst, _fname);

  result = 0;

_exit:
  if (f != NULL) fclose (f);

  DBG_LEAVE (); //EXIT //ins_file_loader_alt
  return result;

_err_fopen:
  *error = "Failed to open input file";
  goto _exit;

_err_fread:
  *error = "Failed to read input file";
  goto _exit;

_err_format:
  *error = "Unknown file format";
  goto _exit;
}

File: iloadins/pas/ins_file_loader_alt.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure ins_file_loader_alt(fname: String);

var
  f: File;
  buffer: tINS_DATA;
  temp: Longint;

function correct_ins(var data): Boolean;

var
  result: Boolean;

begin
  _dbg_enter ({$I %FILE%}, 'ins_file_loader_alt.correct_ins');

  result := TRUE;
  If NOT (tADTRACK2_INS(data).fm_data.WAVEFORM_modulator in [0..3]) then
    result := FALSE;
  If NOT (tADTRACK2_INS(data).fm_data.WAVEFORM_carrier in [0..3]) then
    result := FALSE;
  If NOT (tADTRACK2_INS(data).fm_data.FEEDBACK_FM in [0..15]) then
    result := FALSE;
  correct_ins := result;

  _dbg_leave; //EXIT //ins_file_loader_alt.correct_ins
end;

begin { ins_file_loader_alt }
  _dbg_enter ({$I %FILE%}, 'ins_file_loader_alt');

  FillChar(temp_instrument,SizeOf(temp_instrument),0);
  {$i-}
  Assign(f,fname);
  ResetF(f);
  {$i+}
  If (IOresult <> 0) then
    begin
      CloseF(f);
      _dbg_leave; EXIT; //ins_file_loader_alt
    end;

  If (FileSize(f) > SizeOf(buffer)) then
    begin
      CloseF(f);
      _dbg_leave; EXIT; //ins_file_loader_alt
    end;

  BlockReadF(f,buffer,FileSize(f),temp);
  If (temp <> FileSize(f)) then
    begin
      CloseF(f);
      _dbg_leave; EXIT; //ins_file_loader_alt
    end;

  Case force_ins of
    0: begin
         If (temp = 12) then
           import_standard_instrument_alt(buffer.idata);
         If (temp = 12) and NOT correct_ins(buffer.idata) then
           import_hsc_instrument_alt(buffer.idata)
         else If (temp > 12) then
                import_sat_instrument_alt(buffer.idata);
       end;

    1: import_hsc_instrument_alt(buffer.idata);
    2: import_sat_instrument_alt(buffer.idata);
    3: import_standard_instrument_alt(buffer.idata);
  end;

  CloseF(f);
  load_flag_alt := 1;

  _dbg_leave; //EXIT //ins_file_loader_alt
end;
